ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule.hs

```diff
@@ -558,13 +562,67 @@ splitFragmentAtSlotNo slotNo (fragment :> block) =
 splitFragmentAtSlotNo _ (Empty anchor) =
   (Empty anchor, Empty anchor)
 
+-- | Generate a peer schedule that rolls back repeatedly.
+--
+-- See 'rollbackSpamPointSchedule'.
+rollbackSpamPeer ::
+  SecurityParam ->
+  Bool ->
+  TestFrag ->
+  [NodeState]
+rollbackSpamPeer (SecurityParam sp) bulk frag =
+  concat (take 1 tips ++ zipWith (:) states (drop 1 tips))
+  where
+    tips | bulk = pure <$> drop (k - 1) states
+         | otherwise = take k <$> tails states
+    states = banalStates frag
+    k = fromIntegral sp
+
+-- | Generate a point schedule in which adversaries roll back repeatedly.
+--
+-- Each rollback happens after having sent @k@ blocks, after which the rollback
+-- point is advanced by one block, e.g. with @k = 3@:
+--
+-- 1, 2, 3, 1, 2, 3, 4, 2, 3, 4, 5, 3, 4, 5, 6
+--
+-- When @bulk@ is 'True', the fragments are compressed into a single tick:
+--
+-- 3, 1, 4, 2, 5, 3, 6
+--
+-- HP is reset to the first block of the rollback fragment, which is only necessary
+-- (I think) to avoid @bulk@ mode from degenerating into a linear sequence.
+rollbackSpamPointSchedule ::
+  PointScheduleConfig ->
+  Peers Int ->
+  Bool ->
+  BlockTree TestBlock ->
+  Maybe PointSchedule
+rollbackSpamPointSchedule PointScheduleConfig {psSecurityParam} freqs bulk blockTree =
+  interleaveWithFrequencies freqs peers
```

Comment:
This additionally uses a frequency config, so the generation of the peer schedules and their combination could be separate steps in the API.

Of course this is irrelevant if we want to generate statefully.